common/cuda_hip/distributed/vector_kernels.hpp.inc

```diff
@@ -0,0 +1,97 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void build_local(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const device_matrix_data<ValueType, GlobalIndexType>& input,
+    const distributed::Partition<LocalIndexType, GlobalIndexType>* partition,
+    comm_index_type local_part, matrix::Dense<ValueType>* local_mtx)
+{
+    const auto* range_bounds = partition->get_range_bounds();
+    const auto* range_starting_indices =
+        partition->get_range_starting_indices();
+    const auto* part_ids = partition->get_part_ids();
+    const auto num_ranges = partition->get_num_ranges();
+
+    Array<size_type> range_id{exec, input.get_num_elems()};
+    thrust::upper_bound(thrust::device, range_bounds + 1,
+                        range_bounds + num_ranges + 1,
+                        input.get_const_row_idxs(),
+                        input.get_const_row_idxs() + input.get_num_elems(),
+                        range_id.get_data(), thrust::less<GlobalIndexType>());
+
+    // write values with local rows into the local matrix at the correct index
+    // this needs the following iterators:
+    // - local_row_it: (global_row, range_id) -> local row index
+    // - flat_idx_it: (local_row, col) -> flat index in local matrix values
+    //                                    array
+    // the flat_idx_it is used by the scatter_if as an index map for the values
+    auto map_to_local_row =
+        [range_bounds, range_starting_indices] __host__ __device__(
+            const thrust::tuple<GlobalIndexType, size_type>& idx_range_id)
+        -> LocalIndexType {
```

Comment:
the return type shouldn't be necessary, since you already static_cast to the type and add another LocalIndexType
```suggestion
            const thrust::tuple<GlobalIndexType, size_type>& idx_range_id) {
```